src/client.rs

```diff
@@ -82,6 +82,29 @@ impl<IO> IoSession for TlsStream<IO> {
     }
 }
 
+#[cfg(feature = "early-data")]
+impl<IO> TlsStream<IO>
+where
+    IO: AsyncRead + AsyncWrite + Unpin,
+{
+    fn poll_early_data(&mut self, cx: &mut Context<'_>) {
+        // In the EarlyData state, we have not really established a Tls connection.
+        // Before writing data through `AsyncWrite` and completing the tls handshake,
+        // we ignore read readiness and return to pending.
+        //
+        // In order to avoid event loss,
+        // we need to register a waker and wake it up after tls is connected.
+        if self
+            .early_waker
+            .as_ref()
+            .filter(|waker| cx.waker().will_wake(waker))
+            .is_none()
+        {
+            self.early_waker = Some(cx.waker().clone());
+        }
+    }
+}
+
 impl<IO> AsyncRead for TlsStream<IO>
```

Comment:
I realized that this could also be implement with AsyncBufRead to reduce code duplication

Reply:
Sure, done.